### PR TITLE
Add a read-only TMemFile ctor sharing data.

### PR DESCRIPTION
**Pull request by Axel-Naumann**

### CODE REVIEW

**Comment by pcanal** [comment]
"SharedMem" reads as "shared memory" [ http://man7.org/linux/man-pages/man7/shm_overview.7.html ]

File: io/io/inc/TROSharedMemFile.h [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+#include "TMemFile.h"
+
+#include <memory>
+#include <vector>
+
+class TROSharedMemFile : public TMemFile {
+
+private:
+   std::shared_ptr<const std::vector<char>> fData; // shared file data / content
```
Comment (bluehood) [inline io/io/inc/TROSharedMemFile.h]: the `shared_ptr` could be `const` since copy-assignment is private _and_ deleted. sometimes ptr constness results in compiler optimizations

File: io/io/inc/TROSharedMemFile.h [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+   TROSharedMemFile &operator=(const TROSharedMemFile&) = delete;
+
+public:
+   TROSharedMemFile(const char *name, std::shared_ptr<const std::vector<char>> data);
+   TROSharedMemFile(const TROSharedMemFile &orig);
+
+   void        Print(Option_t *option="") const override;
+
+   ClassDefOverride(TROSharedMemFile, 0) // A ROOT file that reads/writes via HDFS
```
Comment (bluehood) [inline io/io/inc/TROSharedMemFile.h]: via HDFS? :sweat_smile:

File: io/io/src/TROSharedMemFile.cxx [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+*/
+
+#include "TROSharedMemFile.h"
+#include "TError.h"
+
+////////////////////////////////////////////////////////////////////////////////
+/// Usual Constructor.  See the TFile constructor for details.
+
+TROSharedMemFile::TROSharedMemFile(const char *name, std::shared_ptr<const std::vector<char>> data) :
+   TMemFile(name, const_cast<char*>(data->data()), data->size(), "READ", "read-only memfile"),
```
Comment (bluehood) [inline io/io/src/TROSharedMemFile.cxx]: doesn't this copy the buffer inside `TMemFile` (which is what we want to avoid)?

File: io/io/src/TROSharedMemFile.cxx [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+   fData(std::move(data))
+{
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// Copying the content of the TROSharedMemFile into another TROSharedMemFile.
+
+TROSharedMemFile::TROSharedMemFile(const TROSharedMemFile &orig) :
+   TMemFile(orig), fData(orig.fData)
```
Comment (bluehood) [inline io/io/src/TROSharedMemFile.cxx]: this definitely copies the buffer of `orig` inside this TMemFile

File: io/io/test/TROSharedMemFileTests.cxx [hunk 5]
```diff
@@ -0,0 +1,37 @@
+#include "TROSharedMemFile.h"
+
+#include "TError.h"
+#include "TMemFile.h"
+#include "TTree.h"
+
+#include "gtest/gtest.h"
+
+TEST(TFileMerger, CreateWithTFilePointer)
```
Comment (bluehood) [inline io/io/test/TROSharedMemFileTests.cxx]: not sure how, but it would be nice to test that two `TROSharedMemFiles` actually read from a shared memory buffer

**Comment by Axel-Naumann** [comment]
@pcanal it's ready to merge; please review.

File: io/io/test/TROMemFileTests.cxx [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+   TMemFile rosmf("romemfile.root", dataPtr);
+   auto oldIgnoreLevel = gErrorIgnoreLevel;
+   gErrorIgnoreLevel = kBreak;
+   TNamed doNotWrite("doNotWrite", "doNotWrite Title");
+   EXPECT_EQ(0, rosmf.WriteTObject(&doNotWrite));
+   gErrorIgnoreLevel = oldIgnoreLevel;
+}
+
+/// Check that TMemFile uses the original buffer, not a copy (like TMemFile)
```
Comment (pcanal) [inline io/io/test/TROMemFileTests.cxx]: It seems that the comment may have been victim to search&replace idiosyncrasy.  Could you review it?

File: io/io/src/TMemFile.cxx [hunk 7]
```diff
@@ -55,10 +55,20 @@ TMemFile::TMemBlock::TMemBlock() : fPrevious(0), fNext(0), fBuffer(0), fSize(0)
...
+   if (size != -1) {
+      fBuffer = new UChar_t[size];
+      fSize = size;
+   }
 }
 
+////////////////////////////////////////////////////////////////////////////////
+/// Constructor not allocating the memory buffer, for external ownership.
+
+TMemFile::TMemBlock::TMemBlock(UChar_t* data, Long64_t size) :
```
Comment (pcanal) [inline io/io/src/TMemFile.cxx]: Why are both this constructor and the (size==-1) in the previous constructor needed?

File: io/io/inc/TMemFile.h [hunk 8]
```diff
@@ -33,7 +36,8 @@ class TMemFile : public TFile {
       UChar_t   *fBuffer;
       Long64_t   fSize;
    };
-   TMemBlock    fBlockList;   ///< Colletion of memory blocks of size fBlockSize
+   TMemBlock    fBlockList;   ///< Collection of memory blocks of size fgDefaultBlockSize
+   const std::shared_ptr<const std::vector<char>> fExternalData; // shared file data / content
```
Comment (pcanal) [inline io/io/inc/TMemFile.h]: The comment 'prefix' should be ///< to match the others.

Comment (pcanal) [reply io/io/inc/TMemFile.h]: Please use an alias to reduce the typename length so that the members can be aligned.

File: io/io/inc/TMemFile.h [hunk 9]
```diff
@@ -56,11 +60,24 @@ class TMemFile : public TFile {
 
    void ResetObjects(TDirectoryFile *, TFileMergeInfo *) const;
 
+   enum class EMode {
```
Comment (pcanal) [inline io/io/inc/TMemFile.h]: This probably below in TFile.

File: io/io/inc/TMemFile.h [hunk 10]
```diff
@@ -73,7 +90,7 @@ class TMemFile : public TFile {
 
    virtual void        Print(Option_t *option="") const;
 
-   ClassDef(TMemFile, 0) // A ROOT file that reads/writes via HDFS
+   ClassDef(TMemFile, 0) // A ROOT file that reads/writes on a chunk of memory
```
Comment (pcanal) [inline io/io/inc/TMemFile.h]: Thanks :)

File: io/io/src/TMemFile.cxx [hunk 11]
```diff
@@ -188,24 +189,15 @@ TMemFile::TMemFile(const TMemFile &orig) :
...
-      read    = kTRUE;
-      fOption = "READ";
-   }
+   EMode optmode = ParseOption(orig.fOption);
 
    fD = orig.fD; // not really used, so it is okay to have the same value.
    fWritable = orig.fWritable;
 
    // We intentionally allocated just one big buffer for this object.
    orig.CopyTo(fBlockList.fBuffer,fSize);
```
Comment (pcanal) [inline io/io/src/TMemFile.cxx]: Shouldn't we avoid the copy if orig.fExternalData != nullptr?

File: io/io/src/TMemFile.cxx [hunk 12]
```diff
@@ -404,7 +402,7 @@ Int_t TMemFile::SysReadImpl(Int_t, void *buf, Long64_t len)
 {
    TRACE("READ")
 
-   if (fBlockList.fBuffer == 0) {
+   if (fBlockSeek == nullptr || fBlockSeek->fBuffer == nullptr) {
```
Comment (pcanal) [inline io/io/src/TMemFile.cxx]: Why is this change needed?

File: io/io/src/TMemFile.cxx (hunk 7 above)
Comment (Axel-Naumann) [reply io/io/src/TMemFile.cxx]: This one is for external memory that we do not copy, the size==-1 is for memory that we do copy. See the comment "   // size will be -1 when copying an existing buffer into fBuffer."

File: io/io/inc/TMemFile.h (hunk 9 above)
Comment (Axel-Naumann) [reply io/io/inc/TMemFile.h]: Indeed. For future generations :-)

File: io/io/src/TMemFile.cxx (hunk 12 above)
Comment (Axel-Naumann) [reply io/io/src/TMemFile.cxx]: It's a drive-by improvement: `fBlockSeek` and `fBlockSeek->fBuffer` are dereferenced below, so they should be checked (not `fBlockList.fBuffer` which is "contractually connected" to `fBlockSeek->fBuffer`).

File: io/io/src/TMemFile.cxx (hunk 7 above)
Comment (pcanal) [reply io/io/src/TMemFile.cxx]: Then this convention **must** be expressed in the constructor's documentation.

Comment (Axel-Naumann) [reply io/io/src/TMemFile.cxx]: See 3d71859f73cea5ac4e94f087946827a4d3d84716 !
